File: backend/app/api/v1/endpoints/email_automation.py
"""
Email automation endpoints
"""
from fastapi import APIRouter, Depends, HTTPException, status, Query
from typing import List, Optional
from app.middleware.auth import get_current_admin
from app.db.database import get_database
from bson import ObjectId
from datetime import datetime

router = APIRouter()


@router.get("/")
async def get_email_automations(
    organization: Optional[str] = Query(None),
    current_user: dict = Depends(get_current_admin)
):
    """Get all email automations (admin only)"""
    db = await get_database()
    query = {}
    
    if organization:
        query["organization"] = ObjectId(organization)
    
    cursor = db.email_automations.find(query)
    automations = await cursor.to_list(length=100)
    
    result = []
    for automation in automations:
        automation["id"] = str(automation["_id"])
        if automation.get("organization"):
            automation["organization"] = str(automation["organization"])
        del automation["_id"]
        result.append(automation)
    
    return result


@router.get("/{automation_id}")
async def get_email_automation(automation_id: str, current_user: dict = Depends(get_current_admin)):
    """Get email automation by ID (admin only)"""
    db = await get_database()
    
    try:
        automation = await db.email_automations.find_one({"_id": ObjectId(automation_id)})
    except:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Automation not found"
        )
    
    if not automation:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Automation not found"
        )
    
    automation["id"] = str(automation["_id"])
    if automation.get("organization"):
        automation["organization"] = str(automation["organization"])
    del automation["_id"]
    
    return automation


@router.post("/")
async def create_email_automation(automation_data: dict, current_user: dict = Depends(get_current_admin)):
    """Create email automation (admin only)"""
    db = await get_database()
    
    if automation_data.get("organization"):
        automation_data["organization"] = ObjectId(automation_data["organization"])
    
    automation_data["created_at"] = datetime.utcnow()
    
    result = await db.email_automations.insert_one(automation_data)
    automation_data["id"] = str(result.inserted_id)
    if automation_data.get("organization"):
        automation_data["organization"] = str(automation_data["organization"])
    del automation_data["_id"]
    
    return automation_data


@router.put("/{automation_id}")
async def update_email_automation(
    automation_id: str,
    automation_data: dict,
    current_user: dict = Depends(get_current_admin)
):
    """Update email automation (admin only)"""
    db = await get_database()
    
    if automation_data.get("organization"):
        automation_data["organization"] = ObjectId(automation_data["organization"])
    
    await db.email_automations.update_one(
        {"_id": ObjectId(automation_id)},
        {"$set": automation_data}
    )
    
    automation = await db.email_automations.find_one({"_id": ObjectId(automation_id)})
    automation["id"] = str(automation["_id"])
    if automation.get("organization"):
        automation["organization"] = str(automation["organization"])
    del automation["_id"]
    
    return automation


@router.delete("/{automation_id}")
async def delete_email_automation(automation_id: str, current_user: dict = Depends(get_current_admin)):
    """Delete email automation (admin only)"""
    db = await get_database()
    
    await db.email_automations.delete_one({"_id": ObjectId(automation_id)})
    
    return {"message": "Email automation deleted successfully"}


@router.post("/{automation_id}/run")
async def run_email_automation(automation_id: str, current_user: dict = Depends(get_current_admin)):
    """Run email automation manually (admin only)"""
    db = await get_database()
    
    automation = await db.email_automations.find_one({"_id": ObjectId(automation_id)})
    
    if not automation:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Automation not found"
        )
    
    # TODO: Implement automation execution
    return {"message": "Automation executed successfully"}

File: server/routes/emailAutomation.js
/**
 * Email Automation Management Routes
 * Admin Only
 */

import express from 'express'
import EmailAutomation from '../models/EmailAutomation.js'
import { protect, admin } from '../middleware/auth.js'
import { runAutomationManually } from '../workers/emailAutomationWorker.js'

const router = express.Router()

// All routes require admin access
router.use(protect, admin)

/**
 * @route   GET /api/email-automation
 * @desc    Get all email automations
 * @access  Private/Admin
 */
router.get('/', async (req, res) => {
  try {
    const { organization } = req.query
    const query = {}

    if (organization) {
      query.organization = organization
    } else {
      query.$or = [
        { organization: null },
        { organization: { $exists: false } },
      ]
    }

    const automations = await EmailAutomation.find(query)
      .populate('organization', 'name')
      .populate('emailTemplate', 'name')
      .populate('createdBy', 'name email')
      .sort({ createdAt: -1 })

    res.json(automations)
  } catch (error) {
    res.status(500).json({ message: error.message })
  }
})

/**
 * @route   GET /api/email-automation/:id
 * @desc    Get email automation by ID
 * @access  Private/Admin
 */
router.get('/:id', async (req, res) => {
  try {
    const automation = await EmailAutomation.findById(req.params.id)
      .populate('organization', 'name')
      .populate('emailTemplate', 'name subject')
      .populate('createdBy', 'name email')

    if (!automation) {
      return res.status(404).json({ message: 'Automation not found' })
    }

    res.json(automation)
  } catch (error) {
    res.status(500).json({ message: error.message })
  }
})

/**
 * @route   POST /api/email-automation
 * @desc    Create email automation
 * @access  Private/Admin
 */
router.post('/', async (req, res) => {
  try {
    const {
      name,
      type,
      organization,
      isEnabled,
      schedule,
      recipients,
      reportFormat,
      emailTemplate,
    } = req.body

    if (!name || !type) {
      return res.status(400).json({ message: 'Name and type are required' })
    }

    const automationData = {
      name,
      type,
      organization: organization || null,
      isEnabled: isEnabled !== undefined ? isEnabled : true,
      schedule: schedule || {
        time: '09:00',
        timezone: 'UTC',
        dayOfWeek: type === 'weekly-report' ? 1 : null, // Monday
        dayOfMonth: type === 'monthly-report' ? 1 : null, // 1st of month
      },
      recipients: recipients || {
        admins: true,
        organizationManagers: true,
        departmentHeads: true,
        technicians: type === 'daily-open-tickets',
      },
      reportFormat: reportFormat || ['html'],
      emailTemplate: emailTemplate || null,
      createdBy: req.user._id,
    }

    const automation = await EmailAutomation.create(automationData)

    res.status(201).json(automation)
  } catch (error) {
    if (error.code === 11000) {
      return res.status(400).json({ message: 'Automation with this type and organization already exists' })
    }
    res.status(500).json({ message: error.message })
  }
})

/**
 * @route   PUT /api/email-automation/:id
 * @desc    Update email automation
 * @access  Private/Admin
 */
router.put('/:id', async (req, res) => {
  try {
    const {
      name,
      isEnabled,
      schedule,
      recipients,
      reportFormat,
      emailTemplate,
    } = req.body

    const automation = await EmailAutomation.findById(req.params.id)

    if (!automation) {
      return res.status(404).json({ message: 'Automation not found' })
    }

    if (name) automation.name = name
    if (isEnabled !== undefined) automation.isEnabled = isEnabled
    if (schedule) automation.schedule = { ...automation.schedule, ...schedule }
    if (recipients) automation.recipients = { ...automation.recipients, ...recipients }
    if (reportFormat) automation.reportFormat = reportFormat
    if (emailTemplate !== undefined) automation.emailTemplate = emailTemplate

    await automation.save()

    res.json(automation)
  } catch (error) {
    res.status(500).json({ message: error.message })
  }
})

/**
 * @route   DELETE /api/email-automation/:id
 * @desc    Delete email automation
 * @access  Private/Admin
 */
router.delete('/:id', async (req, res) => {
  try {
    const automation = await EmailAutomation.findById(req.params.id)

    if (!automation) {
      return res.status(404).json({ message: 'Automation not found' })
    }

    await automation.deleteOne()

    res.json({ message: 'Automation deleted successfully' })
  } catch (error) {
    res.status(500).json({ message: error.message })
  }
})

/**
 * @route   POST /api/email-automation/:id/run
 * @desc    Run automation manually (for testing)
 * @access  Private/Admin
 */
router.post('/:id/run', async (req, res) => {
  try {
    const result = await runAutomationManually(req.params.id)
    res.json({ success: true, result })
  } catch (error) {
    res.status(500).json({ message: error.message })
  }
})

export default router

